client/src/store.ts

```diff
@@ -0,0 +1,61 @@
+import React, { useReducer, createContext, useContext } from 'react';
+import cloneDeep from 'lodash.clonedeep';
+
+const NODE_ENV = process.env.NODE_ENV || 'development';
+
+type State = {
+  apiUrl: string;
+};
+
+type Action = { type: 'action1' } | { type: 'action2' };
+
+type ActionDictionary = {
+  [key: string]: (state: State, action: Action) => State;
+};
+
+type StoreContext = { state: State; dispatch: React.Dispatch<Action> };
+
+// Define initial application state
+const initialState: State = {
+  apiUrl: {
+    development: 'http://127.0.0.1:8000/api/v1/',
+    production: new URL('/api/v1/', window.location.origin).href,
+    test: 'https://tutv-dev.herokuapp.com/api/v1/',
+  }[NODE_ENV],
+};
+
+// Create the context
+const StoreContext = createContext({} as StoreContext);
+
+// In order to define an action, first define it in the Action type definition above, then add a
+// handler function here.
+// A handler function should take state as input, and return a copy of the state object mutated
+const actionHandlers: ActionDictionary = {
+  action1(state: State): State {
+    console.log('ACTION 1');
+    return state;
+  },
+  action2(state: State): State {
+    console.log('ACTION 2');
+    return state;
+  },
+};
+
+const reducer = (state: State, action: Action): State => {
+  const actionHandler = actionHandlers[action.type];
+  const state2 = cloneDeep(state);
```

Comment:
Why do we need a deep clone?

Reply:
The deep clone is so that action handler functions don't have to be responsible for cloning the state.

If there wasn't a deep clone, you'd normally want action handler functions to go way out of the way to make sure they're returning a copy of the state object rather than the original (since they're supposed to be pure functions that return the new state). This is how an action handler function might do that if it was passed the original (previous) state object:
```js
return { ...state, auth: { ...state.auth }, keyToModify: newValue };
```
You can see that in order to really be thorough in creating a pure new state object that doesn't reference any objects that were part of the previous state, action handler functions have to manually create copies of objects in state that they're not interested in modifying.

When the deep clone is done before the state is passed to the action handlers, the action handler functions can modify the state directly like
```js
state.auth.accessToken = "abcdefg";
```
and not have to worry about mutating the previous state, since the object passed by reference is a brand new state object, and mutating it can't affect the previous state object.

This pattern of passing a pre-cloned state function to the action handler functions is taken from Vuex (see [this](https://vuex.vuejs.org/guide/mutations.html)), but for this case I decided to require returning the state object even after it's mutated just because it's more explicit that way.

Reply:
Hmm I am not familiar with Vuex, but in Redux you are fine with a shallow copy, as long as the outermost reference is new, right? So you could do that with just the spread operator? https://stackoverflow.com/questions/43151622/in-redux-is-it-necessary-to-do-deep-copy

Reply:
I wasn't aware of that distinction, that makes sense. I still think it's nice to be able to modify state imperatively without having to worry about side effects in *any* case, which is what we gain by passing a deep clone, but if you'd rather take the deep clone out I can; let me know

Reply:
I think for this project it should be okay to not have to deep clone. As long as the reference changes, the states are considered unequal and will trigger re-renders.

Reply:
Removed!